### PR TITLE
Audigent RTD Provider HaloId Support & RTD Phase 3 Compliance

### PR DESCRIPTION
**Pull request by anthonylauzon**
## Type of change
- [ ] Bugfix
- [ ] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [X] Other

## Description of change
This PR expands the Audigent RTD provider module to allow support for retrieving Audigent segments for the recently added HaloId among other supported id systems.

## Other information
Related PRs:

Halo Id System Module: https://github.com/prebid/Prebid.js/pull/5524
Halo Id System Documentation: https://github.com/prebid/prebid.github.io/pull/2320
Audigent RTD Provider Module: https://github.com/prebid/Prebid.js/pull/4834

### CODE REVIEW

**Comment by anthonylauzon** [comment]
@pm-harshad-mane this PR is a followup to https://github.com/prebid/Prebid.js/pull/5524 as discussed.

**Comment by bretg** [comment]
- We need to make sure this RTD provider is aligned with the RTD-code changes being made 
in https://github.com/prebid/Prebid.js/pull/5783
- Not seeing docs for this RTD provider. You can base a docs PR off of https://github.com/prebid/prebid.github.io/blob/master/dev-docs/modules/browsiRtdProvider.md

**Comment by anthonylauzon** [comment]
> * https://github.com/prebid/prebid.github.io/blob/master/dev-docs/modules/browsiRtdProvider.md

Documentation has been added:
https://github.com/prebid/prebid.github.io/pull/2369

Are there any detailed breaking code changes in the referenced PR that would require modification here prior to merge?  If not, can we merge this and edit as necessary after any breaking changes in the other open PR are applied?

**Comment by bretg** [comment]
Yes, the `getData` interface is going away, replaced with `getBidRequestData` -- see the docs PR at https://github.com/prebid/prebid.github.io/pull/2268/files

We won't merge this PR or https://github.com/prebid/Prebid.js/pull/5783 until they're all working nicely together. Honestly if I'd been aware of this PR in the first place I wouldn't have approved its release -- the RTD infrastructure is still "beta" until this "phase 3" is done. That said, I'll take the extraordinary step of scheduling a meeting to push RTD review through ASAP, hopefully by next week.

Specifically, your `getSegments(adUnits, onDone)` needs to change to: `getSegments(reqBidsConfigObj, callback, config, userConsent)`

You'll need to read and update the the reqBidsConfigObj, calling the callback with the auctionId when done.

**Comment by anthonylauzon** [comment]
Thanks for the clarification @bretg.  I'll integrate the proposed changes and update this PR.

**Comment by bretg** [comment]
Any progress @anthonylauzon ? We're in the final stages of reviewing the RTD-core changes. If required, we'll make these changes for you:

Minimal changes:
- modify 'getData' to 'getTargetingData'
-  `getSegments(adUnits, onDone)` needs to change to: `getSegments(adUnits, config, userConsent)` and no need to call the callback -- just return data in the proper format.

See https://github.com/prebid/prebid.github.io/pull/2268/files for details

**Comment by anthonylauzon** [comment]
@bretg I have a separate branch tracking https://github.com/prebid/Prebid.js/pull/5783 which I will merge here when that one is pulled into Prebid main.   I was curious about the "getTargetingData" vs 'getBidRequestData' and whether calls to 'getTargetingData' will allow us to augment the ad units with segment data prior to the bid requests being sent.  A cursory reading of the code changes made in 5783 left some uncertainty in this regard.

**Comment by bretg** [comment]
thanks @anthonylauzon , but I was planning on waiting to merge #5783 until this PR and the JW RTD module are ready for it -- don't want to break existing RTD modules. If you would prefer to get this one merged now and open a separate PR to go along with #5783 , guess that would be ok.

> whether calls to 'getTargetingData' will allow us to augment the ad units with segment data prior to the bid requests being sent

Nope, for that you can make a handler to receive getBidRequestData, which can modify anything in the bidRequestObject. Would appreciate any feedback on the docs PR -- https://github.com/prebid/prebid.github.io/pull/2268/files

**Comment by anthonylauzon** [comment]
@bretg I appreciate the feedback. Since the RTD module update isn't backwards compatible, I am concerned about merging untested adaptations to its changing spec prior to the core module's finalization. I will track 5783's branch and be ready to request that this PR be merged as soon as it finally gets through its own PR process.

**Comment by bretg** [comment]
@anthonylauzon - this has become hard to manage -- I'd like to ask that you to update this PR to align with RTD-core phase 3 -- https://github.com/prebid/Prebid.js/pull/5783

Am having trouble getting reviewers who have time at all, let alone reviewers who have time to review it twice. I'll feel better about forcing it on someone internal to my company if it's the new internal interface.

I will make sure that no one merges this before #5783 is merged. That's getting close.

**Comment by anthonylauzon** [comment]
@bretg i've merged my internal rtd3 development branch here.  It should be ready to go, although I would of course like a day to fully test with respect to https://github.com/prebid/Prebid.js/pull/5783 after it is merged.

**Comment by anthonylauzon** [comment]
@bretg this PR should now be ready to merge and meets the https://github.com/prebid/Prebid.js/pull/5783 spec.  None of our publisher partners have implemented it yet, so you are free to merge it before #5783 if you'd like.  The CircleCI build is failing from an unrelated issue or non-deterministic tests.

File: modules/audigentRtdProvider.js
Comment (karimMourra) [inline modules/audigentRtdProvider.js:52]: do you need consent from the user before reading from the local storage ?

File: modules/audigentRtdProvider.js
Comment (karimMourra) [inline modules/audigentRtdProvider.js:68]: you're calling `storage.getDataFromLocalStorage('auHaloId')` twice. I suggest calling it only once, it minifies a lot better. `getDataFromLocalStorage` will not minify and neither will the string `'auHaloId'` so that's a large character count increase. Var names on the other hand minify to 1 character.

```
let haloId = storage.getDataFromLocalStorage('auHaloId');
if (haloId) {
    userIds.haloId = haloId;
    getSegmentsAsync(adUnits, onDone, config, userConsent, userIds);
}
```

Comment (anthonylauzon) [reply modules/audigentRtdProvider.js]: added var names

File: modules/audigentRtdProvider.js
Comment (anthonylauzon) [reply modules/audigentRtdProvider.js]: we're the only one adding to / reading from this variable, so it is only a cache for previously added segments - since the segments would be retrieved from the server here anyways, it's only an optimization

**Comment by anthonylauzon** [comment]
@bretg i've added some minification optimizations per suggestion, but this should now be ready to merge ahead of https://github.com/prebid/Prebid.js/pull/5783.

**Review by karimMourra — APPROVED** [review]

**Pull request merged**
